src/gpu/locks.rs

```diff
@@ -14,31 +14,107 @@ use crate::gpu::{CpuGpuMultiexpKernel, GpuName};
+                Ok(val) if val > 0 => {
+                    let devices = Device::all();
+                    info!(
+                        "BELLPERSON_GPUS_PER_LOCK == {}, try lock {}/{} gpus",
+                        val,
+                        val,
+                        devices.len(),
+                    );
+
+                    let mut locks = Vec::new();
+                    for (index, device) in devices.iter().enumerate() {
+                        let uid = device.unique_id();
+                        let path = tmp_path(GPU_LOCK_NAME, Some(uid));
+                        debug!("Acquiring GPU lock {}/{} at {:?} ...", index, val, &path);
+                        let file = File::create(&path).unwrap_or_else(|_| {
+                            panic!("Cannot create GPU {:?} lock file at {:?}", uid, &path)
+                        });
+                        if file.try_lock_exclusive().is_err() {
+                            continue;
+                        }
+                        debug!("GPU lock acquired at {:?}", path);
+                        locks.push(LockInfo {
+                            file: Some(file),
+                            path: Some(path),
+                            devices: vec![device],
+                        });
+                        if locks.len() >= val {
+                            break;
+                        }
+                    }
+
+                    return GPULock(locks);
+                }
+                Ok(val) if val == 0 => {
+                    info!("BELLPERSON_GPUS_PER_LOCK == 0, free to use gpus");
+                    return GPULock(vec![LockInfo {
+                        file: None,
+                        path: None,
+                        devices: Device::all(),
+                    }]);
+                }
+                _ => warn!("BELLPERSON_GPUS_PER_LOCK parse fail, use all gpus"),
+            };
+        }
+
+        info!("BELLPERSON_GPUS_PER_LOCK fallback to single lock mode");
+
+        // Fallback to create single lock
+        let path = tmp_path(GPU_LOCK_NAME, None);
+        debug!("Acquiring GPU lock at {:?} ...", &path);
+        let file = File::create(&path).unwrap_or_else(|_| {
+            panic!("Cannot create GPU lock file at {:?}", &path);
+        });
+        file.lock_exclusive().unwrap();
         debug!("GPU lock acquired!");
-        GPULock(f)
+        GPULock(vec![LockInfo {
+            file: Some(file),
+            path: Some(path),
+            devices: Device::all(),
+        }])
     }
 }
-impl Drop for GPULock {
+
+impl Drop for GPULock<'_> {
     fn drop(&mut self) {
-        self.0.unlock().unwrap();
-        debug!("GPU lock released!");
+        for lock_info in &self.0 {
+            if let Some(file) = &lock_info.file {
+                file.unlock().unwrap();
+                debug!(
+                    "GPU lock released at {:?}",
+                    lock_info.path.as_ref().unwrap(),
```

Comment:
This `unwrap()` triggered an idea. What if we change the `LockInfo` struct to:

```rust
struct LockInfo<'a> {
    file_path: Option<(File, PathBuf)>,
    devices: Vec<&'a Device>,
}
```

The reason to do it would be, to make it clear that their either is a file together with a path. So it's either both `Some` or `None`.

If we change it, the if statement in this function becomes:

```rust
if let Some((file, path)) = &lock_info.file {
```